modules/nextflow/src/main/groovy/nextflow/file/FilePorter.groovy

```diff
@@ -84,7 +84,20 @@ class FilePorter {
         log.trace "Stage foreign files: $paths"
         def result = new HashMap(filesMap)
 
-        def futures = getExecutor().invokeAll(actions)
+        def stageExecutor = getExecutor()
+        for ( def action : actions ) {
+            stageExecutor.submit(action)
+        }
+
+        def futures = new ArrayList<Future<NamePathPair>>(actions.size())
+        while ( futures.size() < actions.size() ) {
+            def future = stageExecutor.poll(pollTimeout, pollTimeoutUnit)
+            if ( !future ) {
+                log.info1(getDownloadMessage(paths))
+                continue
+            }
+            futures << future
+        }
         log.trace "Stage foreign files completed: $paths"
```

Comment:
What about make this a separate method getting the executor and actions, so that can be tested independently

Reply:
Done